lib/dma.h

```diff
@@ -182,24 +211,33 @@ _dma_mark_dirty(const dma_controller_t *dma, const dma_memory_region_t *region,
     range_to_pages(sg->offset, sg->length, dma->dirty_pgsize,
                    &pgstart, &pgend);
 
-    index = bit_to_u8(pgstart);
-    end = bit_to_u8(pgend) + !!(bit_to_u8off(pgend));
-
-    for (i = index; i < end; i++) {
-        uint8_t bm = ~0;
-
-        /* Mask off any pages in the first u8 that aren't in the range. */
-        if (i == index && bit_to_u8off(pgstart) != 0) {
-            bm &= ~((1 << bit_to_u8off(pgstart)) - 1);
-        }
+    bm_sg_start = bit_to_u8(pgstart);
+    bm_sg_end = bit_to_u8(pgend) + !!(bit_to_u8off(pgend));
+
+    assert(bm_sg_start <= bm_sg_end);
+    
+    u8_set_all(&bm);
+    /* If only one byte of the bitmap has dirty pages, mark them and return. */
+    if (pgstart == pgend || bm_sg_start == bm_sg_end - 1) {
+        mask_any_lower_bits(&bm, pgstart);
+        mask_any_higher_bits(&bm, pgend);
+        apply_byte_mask(&region->dirty_bitmap[bm_sg_start], bm);
+        return;
+    }
 
-        /* Mask off any pages in the last u8 that aren't in the range. */
-        if (i == end - 1 && bit_to_u8off(pgend) != 0) {
-            bm &= ((1 << bit_to_u8off(pgend)) - 1);
-        }
+    /* Mask off any pages in the first u8 that aren't in the range. */
+    mask_any_lower_bits(&bm, pgstart);
+    apply_byte_mask(&region->dirty_bitmap[bm_sg_start], bm);
```

Comment:
can use relaxed consistency now it's split up, and add a barrier at the end

Reply:
ACK RE relaxed. Why do we need the barrier?

Reply:
See previously PR - not sure we do, really, it was paranoia as we had to say something.

Reply:
I assume this would be a smp_wmb()? I'd see no reason for a read barrier since if dma_controller_dirty_page_get() missed the write we would pick it up on the next pass, and concurrent calls to _dma_mark_dirty() won't read it outside of an atomic operation.

The argument that the memset() should be safe for the middle bytes should imply that the order in which the byte writes to the bitmap persist shouldn't matter, right? Why would it be different for the first or last byte, especially considering it is an atomic-or? Therefore I don't think we need a smp_wmb().

> it was paranoia as we had to say something.

I'm struggling to find the discussion. Was there a failure? I'd rather not add barriers we can't fully justify but I'll defer to your judgement.

Reply:
https://github.com/nutanix/libvfio-user/pull/677#discussion_r880374347

So, the question is on the very last pass - is it possible for our writes here not to show up prior to the final grab of bitmap data? I think we might need a full barrier at the point we read the bitmap instead? I haven't thought about it properly, hence the previous seq consistency changes.

> especially considering it is an atomic-or

we still need to pick a consistency flag

Reply:
If you mean smb_wmb() and smb_rmb(), and using relaxed consistency in all atomic ors, then right, that's what I was sugessting. But I also think smb_mb() before dirty_page_Get would work on its own, but again, haven't "done the maths" on that.